Fix transmission config errors

kesquerra opened the pull request:
Add more verbose transmission logging
Update revision number for oresat-live service

AlexanderJDupree commented on tx-rx/dxwifi/tx/tx.c:
```diff
@@ -172,6 +172,59 @@ void delay_transmission(dxwifi_tx_frame* frame, dxwifi_tx_stats stats, void* use
+    float packet_loss_rate = *(float*) user;
+    //generate random num withing range
+    srand((unsigned) time(&t));
+    float random = (srand()%10) / 100.0;
```
@Cam57 I think you mean to use `rand()` here. `srand()` is to seed the generator, `rand()` actually generates the numbers. 
Also, won't this expression result in `random` being in the range (0.0, 0.1) inclusive? Isn't the `packet_loss_rate` domain in the range of (0.0, 1.0)?

Cam57 replied:
Yeah you are correct, the range is between 0.0 and 1.0. I'll check my math and change it.

AlexanderJDupree replied:
I believe you can do 
```c
(double)rand() / (double)RAND_MAX
```
To always give you a number between 0 and 1

AlexanderJDupree commented on tx-rx/dxwifi/tx/tx.c:
```diff
@@ -172,6 +172,59 @@ void delay_transmission(dxwifi_tx_frame* frame, dxwifi_tx_stats stats, void* use
+    int frame_size = DXWIFI_TX_HEADER_SIZE + frame->payload_size + IEEE80211_FCS_SIZE;
+    int total_num_errors = frame_size * error_rate; //Get total number of errors
+    
+    for(int i = 0; i < total_num_errors; ++i){
```
@Cam57 I like what you did here, looping through the total number of expected errors instead of every single byte!  However, I'm confused as to the definition of the bit error rate, and maybe you can clear it up for me. Let's say the bit error rate is 70%, then is it. . . 

A) For every single bit in the frame each bit has a 70% chance of being flipped

or 

B) Out of all the bits in the frame 70% of them will be flipped. 

What implement here is option B, but frame_size is the size of the frame in bytes so the number of errors should actually be

```c
int total_num_errors = frame_size * 8 * error_rate;
```

AlexanderJDupree replied:
If we did decide that option A is correct then it could look something like this: 

```c
    for(size_t i = 0; i < frame_size; ++i) {
        uint8_t mask = 0;

        // Generate a randomized bitmask
        for(size_t bit = 0; bit < 8; ++i) {
            if(((double)rand() / (double)RAND_MAX) < error_rate) {
                mask &= (1 << bit);
            }
        }

        // Apply the mask to the byte, we need to cast to a uint8_t* for the compiler to index the correct byte
        ((uint8_t*)frame)[i] ^= mask;
    }
```

khanu263 replied:
When we first were envisioning the error simulator, option A is definitely what I had in mind. But actually now that I think about it more, option B may be more immediately useful. We're basically going to be using this to see how our FEC performs in the face of worse and worse errors, and being able to flip exactly 40% of the bits every time seems like it would be very convenient functionality to have, as opposed to giving an error rate of 40% and actually getting an error rate of 35%, 46%, 38%, etc.

With that said, once we have pushed the limits of our FEC, it seems like option A would be better for more rapid "realistic" testing.

So I'm inclined to keep things as they are with option B, and we can make implementing option A as a command-line switch an easy low-hanging fruit to pick up in the next sprint...

AlexanderJDupree commented on tx-rx/dxwifi/tx/tx.c:
```diff
@@ -172,6 +172,59 @@ void delay_transmission(dxwifi_tx_frame* frame, dxwifi_tx_stats stats, void* use
+    int total_num_errors = frame_size * error_rate; //Get total number of errors
+    
+    for(int i = 0; i < total_num_errors; ++i){
+        uint8_t chosen_byte = rand()%frame_size;
```
@Cam57 Do we care if the `chosen_byte` and `chosen_bit` has already been chosen? It's theoretically possible to loop through once and flip, let's say bit number 13, and then on another iteration flip bit 13 back to it's original value. Is that possible for a single bit to be flipped multiple times in space? Maybe we should ask @khanu263 or @GavinMegson

khanu263 replied:
If we're going with option B (see my other comment above) then I think this does matter -- if we're flipping 40% of the bits, we want to flip 40% without any "double flips" counting towards the total.

I'd be more OK with it in the context of option A, since it's all random anyway, but if we're enforcing it for option B, I don't see any reason not to for option A either.

AlexanderJDupree replied:
@Cam57 Tracking which bits have already been set could be tricky. I'd recommend using a bit array of length `frame_size` and initialize every value to zero. Then when you generate `chosen_byte` and `chosen_bit` you can first check if the bit has been set in the bit array first. If it hasn't been set, then set the bit in the bit array and toggle the bit in the `frame`.

AlexanderJDupree commented on tx-rx/dxwifi/tx/tx.c:
```diff
@@ -172,6 +172,59 @@ void delay_transmission(dxwifi_tx_frame* frame, dxwifi_tx_stats stats, void* use
+void bit_error_rate_sim(dxwifi_tx_frame* frame, dxwifi_tx_stats stats, void* user) {
+    float error_rate = *(float*) user;
+    srand((unsigned) time(&t)); //init random number 
+    int bit_mask[8] = { //Maybe should be int
```
@Cam57 This is 256 bytes for an 8bit mask. Maybe I'm being nitpicky, but I feel like we can do better. In the loop at around line ~223 you have this:

```c
 int chosen_bit = rand() % 8;
 frame[chosen_byte] ^= bit_mask[chosen_bit];
```
Why not do something like this and drop the bit mask array completely

```c
 int chosen_bit = rand()%8;
 frame[chosen_byte] ^= (1 << chosen_bit);
```

AlexanderJDupree replied:
Also, you're missing a semicolon

Cam57 replied:
I am unsure that bit shifting would be what we want. We would never be able to flip bits in other locations besides bit 1.

AlexanderJDupree replied:
@Cam57 The expression 
```c
1 << chosen_bit
``` 
Will set the bit in the position  of `chosen_bit`. For example, if `chosen_bit` was 3 then left shifting the value `0b00000001`  by three will yield `0b00001000`. So instead of building the bitmask statically and storing the results in a 256 byte array, we can just compute the bits position on the fly.

AlexanderJDupree replied:
In general, the you can do the following operations to manipulate bits: 

```c
// Assume word and k are 32bit integers

// Set the Kth bit
word |= 1 << k;

// Clear the Kth bit
word &= ~(1 << k);

// Toggle the Kth bit
word ^= 1 << k;
```

AlexanderJDupree commented on tx-rx/libdxwifi/details/dirwatch.c:
```diff
@@ -15,7 +15,6 @@
 #include <unistd.h>
 #include <fnmatch.h>
 
-#include <linux/limits.h>
```
@Cam57 this was my bad, I told you to delete this include to test something. But you'll need to add it back in for it to compile.

AlexanderJDupree reviewed (commented):
Requesting some changes. Please see the comments in the review. Thanks!

Cam57 commented on tx-rx/dxwifi/tx/tx.c:
```diff
@@ -172,6 +172,59 @@ void delay_transmission(dxwifi_tx_frame* frame, dxwifi_tx_stats stats, void* use
+void packet_loss_sim(dxwifi_tx_frame* frame, dxwifi_tx_stats stats, void* user) {
+    float packet_loss_rate = *(float*) user;
+    //generate random num withing range
+    srand((unsigned) time(&t));
```
Whoops easy fix!

AlexanderJDupree commented on tx-rx/dxwifi/tx/tx.c:
```diff
@@ -207,24 +207,26 @@ void bit_error_rate_sim(dxwifi_tx_frame* frame, dxwifi_tx_stats stats, void* use
+            --i;
+        }
     }
+    syslog(LOG_INFO, "\nNumber of bit errors: %d", total_num_errors);
```
@Cam57 change this to 

```c
log_info("Number of bit errors: %d", total_num_errors);
```

The dxwifi logging library is capable of switching between syslog and stdout on the fly.

AlexanderJDupree commented on tx-rx/dxwifi/tx/tx.c:
```diff
@@ -207,24 +207,26 @@ void bit_error_rate_sim(dxwifi_tx_frame* frame, dxwifi_tx_stats stats, void* use
     int frame_size = DXWIFI_TX_HEADER_SIZE + frame->payload_size + IEEE80211_FCS_SIZE;
-    int total_num_errors = frame_size * 8 *error_rate; //Get total number of errors
+    int total_num_errors = frame_size * 8 * error_rate; //Get total number of errors
+    int bit_array[frame_size * 8]; //Make an array of bits equal to the number in the frame
```
@Cam57 int is four bytes so here we allocate a stack array of potentially 40,192 bytes,  (I think the default block size is ~1250 bytes), when all we need is a bitset of length `frame_size`. So if we change this line to:

```c
uint8_t bit_array[frame_size];
```

We can get rid of the special addressing and indexing you have going on at line ~221. It would end up looking like this: 

```c
if(bit_array[chosen_byte] & (1 << chosen_bit) == 0) { // Flip bit if unseen
      frame[chosen_byte] ^= (1 << chosen_bit); // Toggle bit in frame
      bit_array[chosen_byte] &= (1 << chosen_bit); // Set bit in the bitset
} 
```

kesquerra committed:
Fix transmission config errors
Add more verbose transmission logging
Update revision number for oresat-live service
